solr/core/src/java/org/apache/solr/cloud/api/collections/Assign.java

```diff
@@ -243,12 +247,24 @@ private static boolean existCoreName(String coreName, Slice slice) {
       }
     } else {
       nodeList = new ArrayList<>(liveNodes);
+      filterNonDataNodes(zk, nodeList);
```

Comment:
In the if branch above (when a node set is passed), the set should likely be filtered as well, and warnings issued if some requested nodes have data role configured to off.